### PR TITLE
dotnet: fix build failing due to running application.

### PR DESCRIPTION
**Pull request by tmds**
### What does this PR do?:

.NET doesn't allow the build to overwrite the files that the running
application is using.

Before starting the build, kill the 'dotnet run' process.

And, add 'UseSharedCompilation=false' to avoid 'dotnet' build processes
that stick around after the 'dotnet build'.

### PR acceptance criteria:

- [x] Contributing guide
_Have you read the [devfile registry contributing guide](https://github.com/devfile/registry/blob/main/CONTRIBUTING.md) and followed its instructions?_
- [ ] Test automation
_Does this repository's tests pass with your changes?_
- [ ] Documentation
_Does any documentation need to be updated with your changes?_

@kadel ptal.

### CODE REVIEW

**Comment by tmds** [comment]
This addresses the following build failure:

```
$ dotnet new web -o web
$ cd web
$ odo create dotnet50
# edit devfile.yaml: update app.csproj to web.csproj
$ odo push
# edit Startup.cs, change `Hello World!` string
$ odo push
...
Executing devfile commands for component dotnet50-web-pdhb
 ✗  Executing build command "dotnet build -c $CONFIGURATION $STARTUP_PROJECT" [16s]
 ✗  Failed to start component with name "dotnet50-web-pdhb". Error: Failed to create the component: command execution failed: unable to execute the run command: unable to exec command [/bin/sh -c cd ${PROJECTS_ROOT} && dotnet build -c $CONFIGURATION $STARTUP_PROJECT]: 
Microsoft (R) Build Engine version 16.9.0+5e4b48a27 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 1 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 2 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 3 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 4 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 5 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 6 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 7 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 8 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 9 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 10 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): error MSB3027: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Exceeded retry count of 10. Failed.  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): error MSB3021: Unable to copy file "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Text file busy [/projects/web.csproj]

Build FAILED.

/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 1 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 2 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 3 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 4 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 5 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 6 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 7 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 8 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 9 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): warning MSB3026: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Beginning retry 10 in 1000ms. Text file busy  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): error MSB3027: Could not copy "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Exceeded retry count of 10. Failed.  [/projects/web.csproj]
/usr/lib64/dotnet/sdk/5.0.208/Microsoft.Common.CurrentVersion.targets(4919,5): error MSB3021: Unable to copy file "/projects/obj/Debug/net5.0/apphost" to "bin/Debug/net5.0/web". Text file busy [/projects/web.csproj]
    10 Warning(s)
    2 Error(s)

Time Elapsed 00:00:14.00
: error while streaming command: command terminated with exit code 1
```
With the change:
```
...
Executing devfile commands for component dotnet50-web-pdhb
 ✓  Executing build command "kill $(pidof dotnet); dotnet build -c $CONFIGURATION $STARTUP_PROJECT /p:UseSharedCompilation=false" [10s]
 ✓  Executing run command "dotnet run -c $CONFIGURATION --no-build --project $STARTUP_PROJECT --no-launch-profile" [2s]

Pushing devfile component "dotnet50-web-pdhb"
 ✓  Changes successfully pushed to component
```

@kadel @maysunfaisal @johnmcollier ptal

File: stacks/dotnet50/devfile.yaml [hunk 1]
```diff
@@ -8,10 +8,10 @@ metadata:
   projectType: dotnet
   tags:
     - dotnet
-  version: 1.0.0
+  version: 1.0.1
 
 starterProjects:
-  - name: dotnet
+  - name: s2i-example
```
Comment (tmds) [inline stacks/dotnet50/devfile.yaml:14]: This is a suggested change unrelated to the fix.

**Comment by kadel** [comment]
> .NET doesn't allow the build to overwrite the files that the running application is using.

Is this related to some certain groups of applications? Asking just out of the curiosity, as I haven't had this problem while I was testing this. But I was only using the s2i example.

**Comment by tmds** [comment]
> Is this related to some certain groups of applications? Asking just out of the curiosity, as I haven't had this problem while I was testing this. But I was only using the s2i example.

It happens for any app that keeps on running, including the web application used by s2i example.

The issue doesn't occur on the first `odo push` because the app is not yet running.
On the second `odo push` you should see the failure.

**Comment by johnmcollier** [comment]
CI Failures:
> FAILURE: FAILED TESTS:  nodejs-nextjs nodejs-nuxtjs nodejs-react nodejs-angular nodejs-vue

Due to known issue: https://github.com/devfile/api/issues/665.

The tests for the dotnet stacks all passed fine, so I have no concerns merging this

**Review by kadel — APPROVED** [review]

**Review by johnmcollier — APPROVED** [review]

**Comment by johnmcollier** [comment]
With the tests for the dotnet tests passing, I'm going to go ahead and merge this.

**Pull request merged**